internal/conf/load.go

```diff
@@ -0,0 +1,196 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package conf
+
+import (
+	"fmt"
+	"github.com/mitchellh/mapstructure"
+	"gopkg.in/yaml.v3"
+	"io/ioutil"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+)
+
+const Separator = "__"
+const Prefix = "KUIPER"
+
+type EnvNameMapping struct {
+	ToCamelCase map[string]string
+	Prefix      string
+}
+
+// These maps are handling case when yaml config option is not lowercase
+
+var KuiperCamelCaseKeyMapping = EnvNameMapping{
+	ToCamelCase: map[string]string{
+		"CONSOLELOG":         "consoleLog",
+		"FILELOG":            "fileLog",
+		"RESTPORT":           "restPort",
+		"RESTTLS":            "restTls",
+		"PROMETHEUSPORT":     "prometheusPort",
+		"PLUGINHOSTS":        "pluginHosts",
+		"CHECKPOINTINTERVAL": "checkpointInterval",
+		"CACHETHRESHOLD":     "cacheThreshold",
+		"CACHETRIGGERCOUNT":  "cacheTriggerCount",
+		"DISABLECACHE":       "disableCache",
+	},
+	Prefix: Prefix,
+}
+
+var MqttSourceCamelCaseKeyMapping = EnvNameMapping{
+	ToCamelCase: map[string]string{
+		"SHAREDSUBSCRIPTION": "sharedSubscription",
+		"CERTIFICATIONPATH":  "certificationPath",
+		"PRIVATEKEYPATH":     "privateKeyPath",
+		"KUBEEDGEVERSION":    "kubeedgeVersion",
+		"KUBEEDGEMODELFILE":  "kubeedgeModelFile",
+	},
+	Prefix: Prefix,
+}
+
+var EdgexCamelCaseKeys = EnvNameMapping{
+	ToCamelCase: map[string]string{
+		"CLIENTID":          "ClientId",
+		"USERNAME":          "Username",
+		"PASSWORD":          "Password",
+		"QOS":               "Qos",
+		"KEEPALIVE":         "KeepAlive",
+		"RETAINED":          "Retained",
+		"CONNECTIONPAYLOAD": "ConnectionPayload",
+		"CERTFILE":          "CertFile",
+		"KEYFILE":           "KeyFile",
+		"CERTPEMBLOCK":      "CertPEMBlock",
+		"KEYPEMBLOCK":       "KeyPEMBlock",
+		"SKIPCERTVERIFY":    "SkipCertVerify",
+		"MESSAGETYPE":       "messageType",
+	},
+	Prefix: Prefix,
+}
+
+var NoCamelCaseMapping = EnvNameMapping{
+	ToCamelCase: map[string]string{},
+	Prefix:      Prefix,
+}
+
+func LoadConfig(c interface{}, mapping EnvNameMapping) error {
+	return LoadConfigByName(ConfFileName, c, mapping)
+}
+
+func LoadConfigByName(name string, c interface{}, mapping EnvNameMapping) error {
+	dir, err := GetConfLoc()
+	if err != nil {
+		return err
+	}
+	p := path.Join(dir, name)
+	return LoadConfigFromPath(p, c, mapping)
+}
+
+func LoadConfigFromPath(path string, c interface{}, mapping EnvNameMapping) error {
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+	configMap := make(map[string]interface{})
+	err = yaml.Unmarshal(b, &configMap)
+	if err != nil {
```

Comment:
I suppose the keys in the configMap are camelCase as they specified in the yaml.  Is the yaml tag in the config struct like  ```ConsoleLog   bool   `yaml:"consoleLog"` ``` is still used in the decode of line 116? For envs, if we let the key to be all lowercase and just set into the configMap, is it possible to let the decode map to the correct case. Maybe it is possible to add a new tag in the config struct field like ```ConsoleLog   bool   `yaml:"consoleLog",json:"consolelog"` ```  in which the json tag name is all lowercase.

Reply:
Is something I committed fine ? I leveraged [this](https://pkg.go.dev/github.com/mitchellh/mapstructure#hdr-Field_Tags)